normalizer_lib.py

```diff
@@ -0,0 +1,158 @@
+
+LANGUAGE_SPECIFIC_NORM = LANGUAGE.LANGUAGE_SPECIFIC_PREPROCESSING
+
+
+# Step 3: Apply NFC unicode normalize
+# really slow when done this way,
+# super fast when I just call unicodedata.normalize
+# on the string at the beginning
+# may remove later
+#def unicode_normalize(string: str) -> Fst:
+#    "Normalizes unicode to NFC normalization."
+#    nfc_normalized = unicodedata.normalize("NFC", string)
+#    normalizer = transducer(string, nfc_normalized).optimize()
+#    return cdrewrite(normalizer, "", "", SIGMA_STAR).optimize()
+
+
+# Step 4: Discard examples not associated with a pronunciation
+# e.g. "how are you today товарищ?" -> "how are you today?"
+
+
+def pass_only_valid_tokens(string: str) -> str:
+    """Replaces invalid tokens in a sentence with <UNK>, keeps the rest.
+
+    Args:
+        string: A line from a corpus.
+
+    Returns:
+        The line, but where invalid tokens have been replaced by <UNK>.
+    """
+    valid_token = INITIAL_PUNCTUATION.ques + Verbalizable.verbalizable() + FINAL_PUNCTUATION.ques
+    returned: List[str] = []
+    remove_extra_whitespace = (string @ REMOVE_EXTRA_WHITESPACE).string()
+    split_string = remove_extra_whitespace.split(" ")
+    for token in split_string:
+        if difference(acceptor(token), valid_token).num_states() == 0:
+            returned.append(token)
+        else:
+            returned.append("<UNK>")
+    return " ".join(returned)
+
+
+# Step 5: Detach punctuation from words
+# e.g. "Who are you?" -> "Who are you ?"
+
+
+NON_GRAPHEME_PUNCT = difference(PUNCTUATION, GRAPHEMES)
+
+INSERT_SPACE = transducer("", SPACE)
+
+SEPARATE_PUNCTUATION = (
+    cdrewrite(
+        INSERT_SPACE,
+        union("[BOS]", SPACE) + union(NON_GRAPHEME_PUNCT),
+        union(PUNCTUATION, Verbalizable.verbalizable()) + union("[EOS]", SPACE, Verbalizable.verbalizable()),
+        SIGMA_STAR) @
+    cdrewrite(
+        INSERT_SPACE,
+        union(Verbalizable.verbalizable(), PUNCTUATION),
+        PUNCTUATION + union("[EOS]", SPACE, PUNCTUATION),
+        SIGMA_STAR))
+
+
+# Step 7: Delete freestanding punctuation
+# e.g. "hi there ?" -> "hi there
+
+DELETE_PUNCTUATION = transducer(PUNCTUATION, "")
+
+DELETE_FREESTANDING_PUNCTUATION = cdrewrite(
+    DELETE_PUNCTUATION,
+    union("[BOS]", SPACE),
+    union("[EOS]", SPACE),
+    SIGMA_STAR)
+
+
+def normalizer(string: str) -> str:
+    """Applies FST rewrite rules to normalize text.
+
+    Args:
+        string: A line from a corpus.
+
+    Returns:
+        The normalized line from the corpus.
+    """
+    string = unicodedata.normalize("NFC", string.lower())
+    #filtered_string = string @ SENTENCE # for all-or-nothing filtering
+    filtered_string = pass_only_valid_tokens(string) # for token-based filtering
+    try:
+        return (filtered_string
+                #@ REMOVE_EXTRA_WHITESPACE
+                @ LANGUAGE_SPECIFIC_NORM
+                #@ unicode_normalize(string) ## SLOW, doing this above instead
```

Comment:
If it's slow maybe just cut this altogether, and the function above.

Reply:
Agreed!